File: src/com/rxjava/CreateOperate.java
package com.rxjava;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

import java.util.concurrent.TimeUnit;

public class CreateOperate {
    public static <T> T getSome() {
        return null;
    }

    public static void main(String[] args) {
        Function function = new Function<Long, Long>() {
            @Override
            public Long apply(Long aLong) throws Exception {
                return aLong + 1000;
            }
        };

        Observable.timer(3, TimeUnit.SECONDS, Schedulers.io())
                .map(function)
                .subscribe(new Observer<Long>() {
            @Override
            public void onSubscribe(Disposable d) {
                System.out.println("onSubscribe--->");
            }

            @Override
            public void onNext(Long s) {
                System.out.println("onNext--->"+s);
                System.out.println("当前线程--->"+Thread.currentThread().getName());
            }

            @Override
            public void onError(Throwable e) {
                System.out.println("onError--->"+e);
            }

            @Override
            public void onComplete() {
                System.out.println("onComplete--->");
            }
        });

        System.out.println("当前线程1--->"+Thread.currentThread().getName());
        try {
            Thread.sleep(10);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        System.out.println("当前线程2--->"+Thread.currentThread().getName());
    }
}

File: src/com/rxjava/CreateOperate.kt
package com.rxjava

import io.reactivex.Observable
import io.reactivex.ObservableOnSubscribe
import io.reactivex.Observer
import io.reactivex.disposables.Disposable
import sun.rmi.runtime.Log
import io.reactivex.schedulers.Schedulers
import java.lang.Thread.sleep
import java.util.concurrent.TimeUnit




private fun testCreate() {
    println("************************")
    println("testCreate()")
    Observable.create(ObservableOnSubscribe<String> { emitter ->
        emitter.onNext("Event1")
        emitter.onNext("Event2")
        emitter.onComplete()
        emitter.onNext("Event3")
    }).subscribe(object : Observer<String> {
        override fun onSubscribe(d: Disposable) {
            println("onSubscribe--->")
        }

        override fun onNext(s: String) {
            println("onNext--->$s")
        }

        override fun onError(e: Throwable) {
            println("onError--->")
        }

        override fun onComplete() {
            println("onComplete--->")
        }
    })

}

internal var lock = Any()
fun testTimer(){
    //timer
    Observable.timer(3, TimeUnit.SECONDS, Schedulers.io()).subscribe(object : Observer<Long> {

        override fun onSubscribe(d: Disposable) {
            println("onSubscribe--->")
        }

        override fun onNext(s: Long) {
            println("onNext--->$s")
            println("当前线程--->" + Thread.currentThread().name)
        }

        override fun onError(e: Throwable) {
            println("onError--->$e")
        }

        override fun onComplete() {
            println("onComplete--->")
        }
    })

    synchronized(lock) {
        try {
            sleep(10)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    println("主线程--->" + Thread.currentThread().name)
}

fun main() {
//    testCreate()
    testTimer()
    sleep(10)

}
